content/x-ray/microservices.md

```diff
@@ -0,0 +1,47 @@
+---
+title: "Deploy Example Microservices"
+date: 2018-11-177T08:30:11-07:00
+weight: 12
+draft: false
+---
+
+We now have the foundation in place to deploy microservices, which are instrumented with [X-Ray SDKs](https://docs.aws.amazon.com/xray/index.html#lang/en_us), to the EKS cluster.
+
+In this step, we are going to deploy example [front-end](https://github.com/rnzsgh/eks-workshop-x-ray-sample-front) and [back-end](https://github.com/rnzsgh/eks-workshop-x-ray-sample-back) microservices to the cluster. The example services are already instrumented using the [X-Ray SDK for Go](https://docs.aws.amazon.com/xray/latest/devguide/xray-sdk-go.html). Currently, X-Ray has SDKs for Go, Python, Node.js, Ruby, .NET and Java.
+
+```
+kubectl apply -f https://raw.githubusercontent.com/rnzsgh/eks-workshop-x-ray-sample-back/master/x-ray-sample-back-k8s.yml
+
+kubectl apply -f https://raw.githubusercontent.com/rnzsgh/eks-workshop-x-ray-sample-front/master/x-ray-sample-front-k8s.yml
+```
+
+To review the status of the deployments, you can run:
+
+```
+kubectl describe deployments x-ray-sample-front-k8s x-ray-sample-back-k8s
+```
+
+For the status of the services, run the following command:
+
+```
+kubectl describe services x-ray-sample-front-k8s x-ray-sample-back-k8s
+```
+
+Once the front-end service is deployed, run the following command to get the Elastic Load Balancer (ELB) endpoint and open it in a browser.
+
+```
+kubectl get service x-ray-sample-front-k8s -o wide
+```
+
+After your ELB is deployed and available, open up the endpoint returned by the previous command in your browser and allow it to remain open. The front-end application makes a new request to the /api endpoint once per second, which in turn calls the back-end service. The JSON document displayed in the browser is the result of the request made to the back-end service.
+
+{{% notice info %}}
+This service was configured with a [LoadBalancer](https://kubernetes.io/docs/tasks/access-application-cluster/create-external-load-balancer/) so,
+an [AWS Elastic Load Balancer](https://aws.amazon.com/elasticloadbalancing/) (ELB) is launched by Kubernetes for the service.
+The EXTERNAL-IP column contains a value that ends with "elb.amazonaws.com" - the full value is the DNS address.
+{{% /notice %}}
+
+{{% notice tip %}}
+When the front-end service is first deployed, it can take up to several minutes for the ELB to be created and DNS updated.
```

Comment:
Is there a command (AWSCLI) I can use to tell when the ELB is ready?

Reply:
There is, but it could be confusing. The ELB may be fully up and running, but DNS may not have propagated yet or may still be cached, so the CLI may return back OK, but still not resolve. Typically, ELBs are allocated really quickly, but DNS can take a bit.

Reply:
Can we add a command or example to show the user how to check/wait until the ELB is ready?

Reply:
I do not have time for this before the testing today.

I worked out the following, but there are edge cases in the kubectl (e.g. - if not initialized yet) command and you need to handle the ELB not being allocated as well. The only way to do this is going to be in a while loop while waiting for multiple conditions to be met. Additionally, I am not sure there is a status field indicating anything, so you I believe you need to look at the length of the array of instances. If you have a better suggestion, please let me now. Even if all of these checks pass, there still isn't a guarantee this will be "ready" because of the additional time it takes for DNS to be updated.

ELB_DNS=$(kubectl get service x-ray-sample-front-k8s -o=custom-columns=:.status.loadBalancer.ingress[0].hostname | sed -n '1!p')
 
ELB_NAME=$(aws elb describe-load-balancers --output text --query "LoadBalancerDescriptions[?DNSName=='$ELB'].LoadBalancerName")

Reply:
let's move this to a separate issue, as it's a "nice to have".